Change orphan hint from "only" to "any uncovered type inside..."

cafce25 opened the pull request:
Fix #128390

cafce25 committed:
Change orphan hint from "only" to "any uncovered type inside..."

cafce25 committed:
Adjust orphan note in tests

cafce25 commented:
@rustbot review

lcnr commented on compiler/rustc_hir_analysis/messages.ftl:
```diff
@@ -351,7 +351,7 @@ hir_analysis_only_current_traits_arbitrary = only traits defined in the current
 hir_analysis_only_current_traits_foreign = this is not defined in the current crate because this is a foreign trait
 
-hir_analysis_only_current_traits_label = impl doesn't use only types from inside the current crate
+hir_analysis_only_current_traits_label = impl doesn't use any uncovered types from inside the current crate
```
this is also subtly wrong

the requirement is that given an impl `impl<T0..TN> Trait<P1..Pn> for P0`, there exists an uncovered local type at position `Pi` such that `P0..Pi` does not contain any uncovered generic parameters

`impl<T> Trait<Local> for T {}` also results in this error

cafce25 replied:
That's correct of course, I didn't want to increase the verbosity too much.

What do you think of appending "before any uncovered generic parameters". This again isn't totally bullet proof as `P0` is only logically before the other `Pn` but is not before them in the source.

lcnr replied:
maybe something like:
- "note: impl has uncovered type parameters before any local type"
- "note: for more information see <link to reference>"

cafce25 replied:
I think it's not so clear that *"no local types"* is also problematic when inverted like thisk. I prefer the original negated version. I do like the shortened *"local type"* instead of *"types from inside the current crate"*, when we link to the reference anyways I think we can get away with it as it expands on what a *"local type"* is: 

- "note: impl doesn't have any local type before any uncovered type parameters"
-  "note: for more information see https://doc.rust-lang.org/reference/items/implementations.html#orphan-rules"

I'm guessing that's the link you meant to add here? Is it ok to link to a fragment of the reference?

I'll see if I can figure out how to add multi line notes (multiple notes?).

lcnr replied:
mutliple notes, yeah

your suggestion looks good to me

spastorino commented:
r? @lcnr (as you're already reviewing it), but this looks good to me with the last suggestion

cafce25 committed:
Add more information link to orphan impls

lcnr commented on tests/ui/coherence/coherence-cow.re_b.stderr:
```diff
@@ -5,7 +5,8 @@ LL | impl<T> Remote for Pair<Cover<T>,T> { }
    | |                  `Pair` is not defined in the current crate
-   | impl doesn't use any uncovered types from inside the current crate
+   | impl doesn't have any local type before any uncovered type parameters
+   | for more information see https://doc.rust-lang.org/reference/items/implementations.html#orphan-rules
```
vibe: I feel like these two lines should be notes, not labels

lcnr commented:
one nit, apart from that r=me

lcnr commented:
going through my list of active reviews today @rustbot author

Dylan-DPC commented:
@cafce25 any updates on this? thanks

cafce25 committed:
Make new information notes instead of labels

cafce25 commented:
@Dylan-DPC Thanks for checking in, dropped the ball a bit here. I've tried to just convert the labels to notes, but that leads to the [notes being mixed with repeats of the span](/cafce25/rust/commit/77cabd908138536d98bc31a63900c7d694c93fe7#diff-e6cf810834f644ec5fdaf3f5c424a37d2e885920e4e641beff4cf603aa74ea34R9-R18) which seems suboptimal so I've also moved them to the `note` field which seems a lot cleaner. @rustbot review

lcnr commented:
@bors r+ rollup

thanks

bors commented:
:pushpin: Commit b89c6204da4331d5d7c181e5c999cd34bfc54e10 has been approved by `lcnr`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=lcnr b89c6204da4331d5d7c181e5c999cd34bfc54e10 -->
<!-- homu: {"type":"Approved","sha":"b89c6204da4331d5d7c181e5c999cd34bfc54e10","approver":"lcnr","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
